crates/sui-indexer-alt/src/handlers/coin_balance_buckets.rs

```diff
@@ -0,0 +1,210 @@
+use sui_indexer_alt_framework::{
+    db,
+    pipeline::{concurrent::Handler, Processor},
+};
+use sui_types::{
+    base_types::{ObjectID, SuiAddress},
+    full_checkpoint_content::CheckpointData,
+    object::{Object, Owner},
+    TypeTag,
+};
+
+use crate::{
+    models::objects::{StoredCoinBalanceBucket, StoredCoinOwnerKind},
+    schema::coin_balance_buckets,
+};
+
+pub(crate) struct CoinBalanceBuckets;
+
+pub(crate) struct ProcessedCoinBalanceBucket {
+    pub object_id: ObjectID,
+    pub cp_sequence_number: u64,
+    pub change: CoinBalanceBucketChangeKind,
+}
+
+pub(crate) enum CoinBalanceBucketChangeKind {
+    Insert {
+        owner_kind: StoredCoinOwnerKind,
+        owner_id: SuiAddress,
+        coin_type: TypeTag,
+        balance_bucket: i16,
+    },
+    Delete,
+}
+
+impl Processor for CoinBalanceBuckets {
+    const NAME: &'static str = "coin_balance_buckets";
+    type Value = ProcessedCoinBalanceBucket;
+
+    // TODO: We need to add tests for this function.
+    fn process(&self, checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
+        let cp_sequence_number = checkpoint.checkpoint_summary.sequence_number;
+        let checkpoint_input_objects = checkpoint.checkpoint_input_objects();
+        let latest_live_output_objects = checkpoint
+            .latest_live_output_objects()
+            .into_iter()
+            .map(|o| (o.id(), o))
+            .collect::<BTreeMap<_, _>>();
+        let mut values: BTreeMap<ObjectID, Self::Value> = BTreeMap::new();
+        for (object_id, input_object) in checkpoint_input_objects.iter() {
+            // This loop processes all coins that were owned by a single address prior to the checkpoint,
+            // but is now deleted or wrapped after the checkpoint.
+            if !input_object.is_coin() {
+                continue;
+            }
+            if get_coin_owner(input_object).is_none() {
+                continue;
+            }
+            if latest_live_output_objects.contains_key(object_id) {
+                continue;
+            }
+            values.insert(
+                *object_id,
+                ProcessedCoinBalanceBucket {
+                    object_id: *object_id,
+                    cp_sequence_number,
+                    change: CoinBalanceBucketChangeKind::Delete,
+                },
+            );
+        }
+        for (object_id, output_object) in latest_live_output_objects.iter() {
+            let Some(coin_type) = output_object.coin_type_maybe() else {
+                continue;
+            };
+
+            let (input_bucket, input_owner) = match checkpoint_input_objects.get(object_id) {
+                Some(input_object) => {
+                    let Some(input_coin) = input_object.as_coin_maybe() else {
+                        bail!("Failed to deserialize Coin for {object_id} in the checkpoint input objects");
+                    };
+                    let bucket = (input_coin.balance.value() as f64).log10() as i16;
```

Comment:
I think this might blow up if the coin has zero balance? Also there's an `ilog10` so better to use that (ditto in the output case).

```suggestion
                    let balance = input_coin.balance.value();
                    let bucket = if balance == 0 {
                        1
                    } else {
                        input_coin.balance.value().ilog10() as i16
                    };
```

Maybe worth factoring this into a helper function like the `get_coin_owner` one.